pandas/core/strings.py

```diff
@@ -372,21 +386,42 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
     0    tWO
     1    bAR
     dtype: object
+
+    Using a compiled regex with flags
+
+    >>> regex_pat = re.compile(r'FUZ', flags=re.IGNORECASE)
+    >>> pd.Series(['foo', 'fuz', np.nan]).str.replace(regex_pat, 'bar')
+    0    foo
+    1    bar
+    2    NaN
+    dtype: object
     """
 
     # Check whether repl is valid (GH 13438, GH 15055)
     if not (is_string_like(repl) or callable(repl)):
         raise TypeError("repl must be a string or callable")
-    use_re = not case or len(pat) > 1 or flags or callable(repl)
 
-    if use_re:
-        if not case:
+    is_compiled_re = is_re(pat)
+    if is_compiled_re:
+        if (case is not None) or (flags != 0):
+            raise ValueError("case and flags must be default values"
+                             " when pat is a compiled regex")
+    else:
+        # not a compiled regex
+        # set default case
+        if case is None:
+            case = True
+
+        # add case flag, if provided
+        if case is False:
             flags |= re.IGNORECASE
-        regex = re.compile(pat, flags=flags)
```

Comment:
You're definitely onto something that I overlooked: `re.sub` is slower than using `<compiled_regex>.sub`. This is because `re.sub` calls `re._compile`, which performs a cache lookup of already-cached regex or compiles the regex (see [python/cpython/Lib/re.py](https://github.com/python/cpython/blob/master/Lib/re.py#L289)), whereas the `<compiled_regex>.sub` does not. 

I will change the code so that we avoid using `re.sub` and avoid the performance hit of the checking the cache each time.